phpBB/phpbb/ucp/controller/reset_password.php

```diff
@@ -0,0 +1,436 @@
+
+				if (!$user_row)
+				{
+					return $this->helper->message($message);
+				}
+
+				if ($user_row['user_type'] == USER_IGNORE || $user_row['user_type'] == USER_INACTIVE)
+				{
+					return $this->helper->message($message);
+				}
+
+				// Do not create multiple valid reset tokens
+				if (!empty($user_row['reset_token']) && (int) $user_row['reset_token_expiration'] >= time())
+				{
+					return $this->helper->message($message);
+				}
+
+				// Check users permissions
+				$auth2 = new auth();
+				$auth2->acl($user_row);
+
+				if (!$auth2->acl_get('u_chgpasswd'))
+				{
+					return $this->helper->message($message);
+				}
+
+				// Generate reset token
+				$reset_token = strtolower(gen_rand_string(32));
+
+				$sql_ary = [
+					'reset_token'				=> $reset_token,
+					'reset_token_expiration'	=> strtotime('+1 day'),
+				];
+
+				$sql = 'UPDATE ' . $this->tables['users'] . '
+					SET ' . $this->db->sql_build_array('UPDATE', $sql_ary) . '
+					WHERE user_id = ' . $user_row['user_id'];
+				$this->db->sql_query($sql);
+
+				if (!class_exists('messenger'))
+				{
+					include($this->root_path . 'includes/functions_messenger.' . $this->php_ext);
+				}
+
+				/** @var \messenger $messenger */
+				$messenger = new \messenger(false);
+
+				$messenger->template('user_forgot_password', $user_row['user_lang']);
+
+				$messenger->set_addresses($user_row);
+
+				$messenger->anti_abuse_headers($this->config, $this->user);
+
+				$messenger->assign_vars([
+						'USERNAME'			=> htmlspecialchars_decode($user_row['username']),
+						'U_RESET_PASSWORD'	=> generate_board_url(true) . $this->helper->route('phpbb_ucp_reset_password_controller', [
+							'u'		=> $user_row['user_id'],
+							'token'	=> $reset_token,
+						], false)
+				]);
+
+				$messenger->send($user_row['user_notify_type']);
+
+				return $this->helper->message($message);
+			}
+		}
+
+		$this->template->assign_vars([
+			'USERNAME'			=> $username,
+			'EMAIL'				=> $email,
+			'S_PROFILE_ACTION'	=> $this->helper->route('phpbb_ucp_forgot_password_controller'),
+		]);
+
+		return $this->helper->render('ucp_reset_password.html', $this->language->lang('RESET_PASSWORD'));
+	}
+
+	/**
+	 * Handle controller requests
+	 *
+	 * @return Response
+	 */
+	public function reset()
+	{
+		$this->init_controller();
+
+		$submit			= $this->request->is_set_post('submit');
+		$reset_token	= $this->request->variable('token', '');
+		$user_id		= $this->request->variable('u', 0);
+
+		if (empty($reset_token))
+		{
+			return $this->helper->message('NO_RESET_TOKEN');
+		}
+
+		if (!$user_id)
+		{
+			return $this->helper->message('NO_USER');
+		}
+
+		add_form_key('ucp_reset_password');
+
+		$sql_array = [
+			'SELECT'	=> 'user_id, username, user_permissions, user_email, user_jabber, user_notify_type, user_type,'
+				. ' user_lang, user_inactive_reason, reset_token, reset_token_expiration',
+			'FROM'		=> [$this->tables['users'] => 'u'],
+			'WHERE'		=> 'user_id = ' . $user_id,
```

Comment:
Eventhough it's already typecasted by `$request`, perhaps here again to keep SQL queries consistent?